ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceTest.java

```diff
@@ -984,6 +968,47 @@ private void doExternalServicesBadInputTest(RestMethod[] restMethods, String exp
   // routerExceptionPipelineTest() helpers.
 
   /**
+   * Does the exception pipelining test for {@link Router}.
+   * @param testRouter the {@link Router} to use to while creating {@link AmbryBlobStorageService}.
+   * @param exceptionMsg the expected exception message.
+   * @throws Exception
+   */
+  private void doRouterExceptionPipelineTest(FrontendTestRouter testRouter, String exceptionMsg)
+      throws Exception {
+    ambryBlobStorageService =
+        new AmbryBlobStorageService(frontendConfig, frontendMetrics, responseHandler, testRouter, idConverterFactory,
+            securityServiceFactory);
+    ambryBlobStorageService.start();
+    for (RestMethod restMethod : RestMethod.values()) {
+      switch (restMethod) {
+        case HEAD:
+          testRouter.exceptionOpType = FrontendTestRouter.OpType.GetBlobInfo;
+          checkRouterExceptionPipeline(exceptionMsg, createRestRequest(restMethod, "/", null, null));
+          break;
+        case GET:
+          testRouter.exceptionOpType = FrontendTestRouter.OpType.GetBlobInfo;
+          checkRouterExceptionPipeline(exceptionMsg, createRestRequest(restMethod, "/", null, null));
+          testRouter.exceptionOpType = FrontendTestRouter.OpType.GetBlob;
```

Comment:
Not sure I understand this comment. Why would testing with different sub-resources be useful here? The test is testing for reaction to router API exceptions. There isn't a test for the `BlobInfo` sub-resource either, so why the question about `UserMetadata`?